Use show by default for generating links.

VincentLanglet opened the pull request:
## Subject

I took a look at issue #3936 and discovered multiples things
1) `addIdentifier` is using `edit` if possible, then `show`.
2) the `fieldDescriptionFactory::create` method of `ORM` use `show` as default.
3) the `fieldDescriptionFactory::create` method of `ODM` use `edit` as default.
4) the twig templates doesn't prevent themself against the fact the option might not be set
5) the searchResultAction is using `edit` if possible, then `show`.

Instead of setting the value everywhere, we should just use a default value in the template.
By solving 4, we avoid the need of 1, 2 and 3.

Which value ? Best would be to provide some configuration/extension points like the default value to use in the twig.
But until we got some feature request, I propose an opinionated decision: 
- When looking for a result in the search page, we first want to look at the description of this result. So the `show` route seems to have more sens.
- When on the list, I click on the identifier url of a oTo or oTm relation, I prefer to first look at the description.
The general rule would be "I prefer see the informations of an object before editing it".

So I added consistency by setting `show` by default everywhere.
Do you agree @sonata-project/contributors ? 

Then, the `final` method `getSearchResultLink` went pretty useless and could be implemented directly in the template. It means less logic in the AbstractAdmin, and later, if someone want to change the way the url is generated, we should just provide a `{% block search_result %}` in the template, instead of a method in the AdminInterface.

Closes #3936.

After the PR, I'll open a PR on persistence bundle, to remove the default value set here
https://github.com/sonata-project/SonataDoctrineORMAdminBundle/blob/4.x/src/FieldDescription/FieldDescriptionFactory.php#L36-L42
https://github.com/sonata-project/SonataDoctrineMongoDBAdminBundle/blob/4.x/src/FieldDescription/FieldDescriptionFactory.php#L36-L42
To rely on the default value of the AdminBundle.

I target this branch because currently we have a inconsistency in the default route used in Sonata 4, especially from one persistence bundle to another ; and it's last time for small BC-break.

## Changelog

<!-- REMOVE EMPTY SECTIONS -->
```markdown
### Changed
- The route used by search result is now `show`.
- The route used by `ListMapper::addIdentifier()` is now `show`

### Removed
- `AbstractAdmin::searchResultAction()`
- `AbstractAdmin::getSearchResultLink()`
- `AdminInterface::getSearchResultLink()`
```

core23 commented on src/Resources/views/Block/block_search_result.html.twig:
```diff
@@ -59,10 +59,11 @@ file that was distributed with this source code.
                         {% for result in current_page_results %}
-                            {% set link = admin.getSearchResultLink(result) %}
-                            {% if link %}
+                            {% if admin.hasRoute('show') and admin.hasAccess('show', result) %}
```
If I got you right, with this PR the show route will be the default route for every search result item? So if you never define any fields for show form, you can never *view* the item details?

If this is the case, I'm 👎

core23 replied:
> * the `fieldDescriptionFactory::create` method of `ORM` use `show` as default.
> * the `fieldDescriptionFactory::create` method of `ODM` use `edit` as default.

I know that there are some inconsistencies, but we should fix them instead of removing a big feature.

VincentLanglet replied:
> You can never _view_ the item details?

If you don't have fields for show form, you already don't view the item details.

And you already cannot view the item details when clicking on a link in the list of an admin since the route used is also, and already, the `show` one.

VincentLanglet replied:
> > * the `fieldDescriptionFactory::create` method of `ORM` use `show` as default.
> > * the `fieldDescriptionFactory::create` method of `ODM` use `edit` as default.
> 
> I know that there are some inconsistencies, but we should fix them instead of removing a big feature.
> 
> Maybe we should rename the method `getSearchResultLink` to `getDetaultItemLink` and use it everywhere

I won't call this "a big feature"
- There was multiple issue where people asked for using the `show` instead of the `edit` ; or at least configuring it.
- There is currently placed where we use `edit` if possible and `show` elsewhere ; and place where one of them is hardcoded, creating more mess and annoyance than anything else.

And I won't say the feature is removed.
In the same way you want to remove the formTheme from the admin, this logic is much better in the template than in the admin class. So people should just override the template. As I said, a `search_result` block can be added to make this easier.

VincentLanglet replied:
I added a `search_result` block for an easy override.

jordisala1991 commented:
> When on the list, I click on the identifier url of a oTo or oTm relation, I prefer to first look at the description.
The general rule would be "I prefer see the informations of an object before editing it".
So I added consistency by setting show by default everywhere.
Do you agree @sonata-project/contributors ?

I do not. I use the addIdentifier to have an easy way to route to edit without adding the edit button. IMO if we are doing this change it should be atleast configurable if I want to go to edit by default everywhere or show.

I personally do not use show at all. Normally disable that route on all admins.

dmaicher commented:
> I personally do not use show at all. Normally disable that route on all admins.

same for me

VincentLanglet commented:
Ok, so after more investigation, I found other situation where a `default_route` was needed/used.

Currently all our usage are
1) In the `addIdentifier`
2) In the twig template of all the OTO/OTM/... fields
3) In the FieldDescriptionFactory (of persistence bundle)
4) In the search results
5) In the breadcrumb when using a child admin

```
IMO if we are doing this change it should be atleast configurable if I want to go to edit by default everywhere or show.
```

For the situation 5, there was already a `breadcrumb.child_admin_route`.
For the situtation 4, I introduced a `global_search.admin_route`.

I plan to remove the occurence 3, (and did remove the 1) to rely only on 2 for which I introduced a `default_admin_route` config option.

With this PR,
- the same route is used everywhere.
- all the logic is in the AdminBundle and can be removed from the persistence bundle
- there are option to change every route used
````
sonata_admin:
    options:
        default_admin_route: edit
    global_search:
        admin_route: edit
    breadcrumbs:
        child_admin_route: edit
```

VincentLanglet commented on src/Block/AdminSearchBlockService.php:
```diff
@@ -105,6 +113,7 @@ public function execute(BlockContextInterface $blockContext, ?Response $response
             'filters' => $filters,
             'admin' => $admin,
             'show_empty_boxes' => $this->emptyBoxesOption,
+            'default_admin_route' => $this->defaultAdminRoute,
```
Indeed, fixed

VincentLanglet commented:
The new options added should solve your issues @core23

SonataCI commented:
Could you please rebase your PR and fix merge conflicts?

SonataCI commented:
Could you please rebase your PR and fix merge conflicts?

jordisala1991 commented:
Thanks @VincentLanglet
